src/lib.rs

```diff
@@ -1653,6 +1653,96 @@ impl Bindings {
+            },
+            Ok(None) => (),
+            Err(err) => println!("{:?}", err),
+        }
+        
+        Ok(())
+    }
+    
+    /// Checks if rustfmt_bindings is set and runs rustfmt on the string
+    fn rustfmt_generated_string(&self, source: String) -> io::Result<Option<(String, String)>> {
+        let _t = self.context.timer("rustfmt_generated_string");
+
+        if !self.context.options().rustfmt_bindings {
+            return Ok(None);
+        }
+
+        let rustfmt = if let Ok(rustfmt) = which::which("rustfmt") {
+            rustfmt
+        } else {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Rustfmt activated, but it could not be found in global path.",
+            ));
+        };
+        
+        let mut cmd = Command::new(rustfmt);
+        cmd
+            .args(&["--write-mode=display"])
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped());
+            
+        if let Some(path) = self.context
+            .options()
+            .rustfmt_configuration_file
+            .as_ref()
+            .and_then(|f| f.to_str())
+        {
+            cmd.args(&["--config-path", path]);
+        }
+        
+        let mut child = cmd.spawn().expect("Failed to spawn rustfmt.");
+        let mut child_stdin = child.stdin.take().unwrap();
+        let mut child_stdout = child.stdout.take().unwrap();
+        let mut child_stderr = child.stderr.take().unwrap();
+    
+        // Write to stdin in a new thread, so that we can read from stdout on this
+        // thread. This keeps the child from blocking on writing to its stdout which
+        // might block us from writing to its stdin.
+        let stdin_handle = ::std::thread::spawn(move || {
+            child_stdin.write_all(source.as_bytes())
+        });
+    
+        // Read stderr on a new thread for similar reasons.
+        let stderr_handle = ::std::thread::spawn(move || {
+          let mut output = vec![];
+          io::copy(&mut child_stderr, &mut output)
+              .map(|_| String::from_utf8_lossy(&output).to_string())
+        });
+
+        let mut output = vec![];
+        io::copy(&mut child_stdout, &mut output)
+            .expect("Should copy stdout into vec OK");
+            
+        // Ignore actual rustfmt status because it is often non-zero for trivial
+        // things.
+        let _ = child.wait().expect("should wait on rustfmt child OK");
+    
+        stdin_handle.join()
+            .expect("writer thread should not have panicked")
+            .expect("should have written to child rustfmt's stdin OK");
+    
+        let bindings = String::from_utf8(output)
+          .expect("rustfmt should only emit valid utf-8");
+    
+        let child_stderr = stderr_handle.join()
+          .expect("stderr reader thread should not have panicked")
+          .expect("should have read child rustfmt's stderr OK");
```

Comment:
All these `expect`s in this function should be propagating errors instead of unwrapping them.

Reply:
Thank you for taking the time to give me thorough feedback @fitzgen.

I will make the relevant changes this evening.

As an aside, this continues to be a great exercise for a crash course through rust! Thank you.